parser_library/src/parsing/parser_impl.cpp

```diff
@@ -14,453 +14,4602 @@
 
-    auto startId = start->getTokenIndex();
-    auto stopId = stop->getTokenIndex();
+    if (auto [error, op] = asm_op(); error)
+    {
+        errors = true;
+        operands.push_back(std::make_unique<semantics::empty_operand>(empty_range(op_start)));
+        while (except<U',', U' '>())
+            consume();
+    }
+    else
+        operands.push_back(std::move(op));
 
-    for (auto id = startId; id <= stopId; ++id)
-        if (auto token = input.get(id);
-            token->getChannel() == lexing::lexer::Channels::DEFAULT_CHANNEL && token->getType() != antlr4::Token::EOF)
-            s.append(token->getText());
+    while (follows<U','>())
+    {
+        consume();
+        op_start = cur_pos_adjusted();
+        if (follows<U',', U' '>())
+        {
+            operands.push_back(std::make_unique<semantics::empty_operand>(empty_range(op_start)));
+            continue;
+        }
+        auto [error, op] = asm_op();
+        if (error || errors)
+        {
+            errors = true;
+            operands.push_back(std::make_unique<semantics::empty_operand>(empty_range(op_start)));
+            while (except<U',', U' '>())
+                consume();
+        }
+        else
+            operands.push_back(std::move(op));
+    }
+    const range r = range_from(start);
+    holder->collector.set_operand_remark_field(std::move(operands), std::vector<range>(), r);
 }
 
-bool parser_impl::goff() const noexcept { return hlasm_ctx->goff(); }
+void parser_holder::lookahead_operands_and_remarks_asm()
+{
+    parser2 p(this);
 
-parser_holder::~parser_holder() = default;
+    p.lookahead_operands_and_remarks_asm();
+}
 
-void parser_holder::prepare_parser(lexing::u8string_view_with_newlines text,
-    context::hlasm_context* hlasm_ctx,
-    diagnostic_op_consumer* diags,
-    semantics::range_provider range_prov,
-    range text_range,
-    size_t logical_column,
-    const processing::processing_status& proc_status) const
+semantics::literal_si parser_holder::literal_reparse()
 {
-    lex->reset(text, text_range.start, logical_column);
-
-    stream->reset();
+    parser2 p(this);
 
-    parser->reinitialize(hlasm_ctx, std::move(range_prov), proc_status, diags);
+    auto [error, lit] = p.lex_literal();
+    if (error || !p.eof())
+        return nullptr;
+    return std::move(lit);
+}
 
-    parser->reset();
+expressions::ca_expr_ptr parser_holder::testing_expr()
+{
+    parser2 p(this);
+    auto [error, e] = p.lex_expr();
+    if (error)
+    {
+        p.add_diagnostic(diagnostic_op::error_S0002);
+        return {};
+    }
+    else
+        return std::move(e);
+}
 
-    parser->get_collector().prepare_for_next_statement();
+expressions::data_definition parser_holder::testing_data_def()
+{
+    parser2 p(this);
+    auto [error, dd] = p.lex_data_definition(false);
+    if (error)
+    {
+        p.add_diagnostic(diagnostic_op::error_S0002);
+        return {};
+    }
+    else
+        return std::move(dd);
```

Comment:
I think there doesn't need to be `std::move` here, right? The same goes for the two function right above this one. Also for `macro_ops()`, `operand_mach()` and line 4168.

Reply:
The automatic move does not happen when returning a "variable" defined in a structured binding.

Reply:
I thought that it might be the case. Well... Thx for letting me know!